sw/host/opentitanlib/src/transport/qemu/subprocess.rs

```diff
@@ -0,0 +1,235 @@
+    pub timeout: Duration,
+}
+
+/// This struct is responsible for managing a QEMU subprocess launched here.
+pub struct Subprocess {
+    /// QEMU subprocess.
+    child: Child,
+
+    /// Directory where sockets connected to QEMU will be stored.
+    ///
+    /// Removed when dropped.
+    socket_dir: TempDir,
+
+    /// Accumulated stderr output from QEMU itself.
+    pub accumulated_stderr: Arc<Mutex<String>>,
+}
+
+impl Subprocess {
+    const UART0_SOCKET: &str = "uart0.socket";
+
+    /// Starts a QEMU [`Subprocess`] based on [`Options`].
+    pub fn from_options(options: Options) -> anyhow::Result<Self> {
+        let mut command = Command::new(&options.executable);
+
+        // Join the machine name with machine properties by commas.
+        let machine = options
+            .machine_props
+            .iter()
+            .fold(options.machine, |mut s, prop| {
+                s.push(',');
+                s.push_str(prop);
+                s
+            });
+        command.args(["-machine", &machine]);
+
+        // Add arguments for specified images:
+
+        if let Some(rom_image) = options.rom_image {
+            command.arg("-kernel");
+            command.arg(rom_image);
+        }
+
+        // NOTE: We load these drives as `readonly` because Bazel can
+        // only provide readonly test inputs. To support tests which modify
+        // flash or OTP, we will need to make writable copies of the images
+        // ourselves.
+
+        if let Some(flash_image) = options.flash_image {
+            command.arg("-drive");
+            command.arg(format!(
+                "if=mtd,bus=1,file={},format=raw,readonly=on",
+                flash_image.display()
+            ));
+        }
+
+        if let Some(otp_image) = options.otp_image {
+            command.arg("-drive");
+            command.arg(format!(
+                "if=pflash,file={},format=raw,readonly=on",
+                otp_image.display()
+            ));
+        }
+
+        // Create files for QEMU to connect IO to.
+        let socket_dir = tempfile::tempdir().context("failed to create tempdir")?;
+
+        let uart_path = socket_dir.path().join(Self::UART0_SOCKET);
+        command.arg("-chardev");
+        command.arg(format!(
+            "socket,id=uart0,server=on,path={}",
+            uart_path.display()
+        ));
+        command.args(["-serial", "chardev:uart0"]);
+
+        // Add explicitly requested arguments.
+        command.args(options.extra_args.as_slice());
+
+        log::info!("CWD: {:?}", env::current_dir());
+        log::info!(
+            "Spawning QEMU: {} {}",
+            options.executable.display(),
+            command
+                .get_args()
+                .map(|s| s.to_string_lossy())
+                .collect::<Vec<_>>()
+                .join(" ")
+        );
+
+        // Start QEMU.
+        let mut child = command
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
```

Comment:
What happens to the stdout? Shouldn't it be either piped and printed, or nulled?

Reply:
I'll null it since I don't expect to use QEMU's stdout for any of the sockets

Reply:
Although now I see that one of the unit tests is wanting to read from `stdout` so maybe I'll accumulate it in the same place as the stderr.